### PR TITLE
PHP 8: Code Review `Survey`

### PR DESCRIPTION
**Pull request by mjansenDatabay**
Hi @alex40724,

I completed the review of the `Modules/Survey` component. I'll provide a separate PR for the `SurveyQuestionPool`.

This component looked very good, but of course I found some issues.

Results:
- [ ] The code style is `PSR-2 + X` compliant: 1 File has been changed by the `PHP-CS-FIXER`
- [x] No usage of `$_GET` / `$_POST` / `$_REQUEST` / `$_SESSION`
- [x] There is a `Unit Test Suite` with at least one unit test
- [x] The unit tests pass
- [x] External libraries are compatible with PHP 8 (if relevant)
- [x] There are no serious `PhpStorm Code Inspection` issues left
- [ ] The types are fully documented (PhpDoc) or explict PHP types are used (type hints, return types, typed properties)

I also fixed some trivial issues reported by my inspection profile.

--

Actions needed:
- [ ] Please check my inline comments. Some of them are just information, some require an action. You can use `TODO PHP8-REVIEW` when searching.

Best regards,
@mjansenDatabay

### CODE REVIEW

File: Modules/Survey/Service/class.InternalGUIService.php [hunk 1]
```diff
@@ -30,7 +30,6 @@ class InternalGUIService
 {
     use GlobalDICGUIServices;
 
-    protected \ilObjectServiceInterface $object_service;
```
Comment (mjansenDatabay) [inline Modules/Survey/Service/class.InternalGUIService.php:33]: This is already provided by the `GlobalDICGUIServices` trait.

File: Modules/Survey/Participants/class.ilSpecialUsersTableGUI.php [hunk 2]
```diff
@@ -46,7 +46,7 @@ public function __construct(
     protected function fillRow(array $a_set) : void
     {
         $ilCtrl = $this->ctrl;
-        $user = ilObjUser::_lookupFields($a_set);
+        $user = ilObjUser::_lookupFields($a_set); // This cannot work
```
Comment (mjansenDatabay) [inline Modules/Survey/Participants/class.ilSpecialUsersTableGUI.php:49]: An `array` is passed here, an `int` is expected.

File: Modules/Survey/Mode/IndividualFeedback/class.UIModifier.php [hunk 3]
```diff
@@ -122,6 +124,7 @@ public function setResultsDetailToolbar(
         $this->addExportAndPrintButton(
             $toolbar,
             true
+            // A parameter is missing here
```
Comment (mjansenDatabay) [inline Modules/Survey/Mode/IndividualFeedback/class.UIModifier.php:127]: Please provide the correct argument here

File: Modules/Survey/Export/class.ilSurveyDataSet.php [hunk 4]
```diff
@@ -130,15 +130,15 @@ public function importRecord(
...
+                    $skill_survey = new ilSurveySkill($this->getImport()->getSurvey()); // This does not work
                     $skill_survey->addQuestionSkillAssignment($q_id, $skill_data[0]["skill_id"], $skill_data[0]["tref_id"]);
                 }
                 break;
 
             case "svy_skill_threshold":
                 $l = ilBasicSkill::getLevelIdForImportIdMatchSkill($this->getCurrentInstallationId(), $a_rec["LevelId"], $a_rec["BaseSkillId"], $a_rec["TrefId"]);
                 if (count($l) > 0) {
-                    $skill_thres = new ilSurveySkillThresholds($this->getImport()->getSurvey());
+                    $skill_thres = new ilSurveySkillThresholds($this->getImport()->getSurvey()); // This does not work
```
Comment (mjansenDatabay) [inline Modules/Survey/Export/class.ilSurveyDataSet.php:141]: `getSurvey` does not exist

File: Modules/Survey/Editing/class.ilSurveyEditorGUI.php [hunk 5]
```diff
@@ -663,11 +664,11 @@ public function executeCreateQuestionObject($q_type = null, $pool_usage = null,
             $obj_id = $this->object->getId();
         }
         // existing pool
-        elseif ($pool_usage == 3 && strlen($this->request->getSelectedPool())) {
+        elseif ($pool_usage == 3 && strlen($this->request->getSelectedPool())) { // TODO An int is returned and used for strlen
```
Comment (mjansenDatabay) [inline Modules/Survey/Editing/class.ilSurveyEditorGUI.php:667]: An int is returned and used for strlen

File: Modules/Survey/Editing/class.ilSurveyEditorGUI.php [hunk 6]
```diff
@@ -663,11 +664,11 @@ public function executeCreateQuestionObject($q_type = null, $pool_usage = null,
...
             $obj_id = ilObject::_lookupObjId($this->request->getSelectedPool());
         }
         // new pool
-        elseif ($pool_usage == 2 && strlen($this->request->getPoolName())) {
+        elseif ($pool_usage == 2 && strlen($this->request->getPoolName())) { // TODO An int is returned and used for strlen
```
Comment (mjansenDatabay) [inline Modules/Survey/Editing/class.ilSurveyEditorGUI.php:671]: An int is returned and used for strlen

File: Modules/Survey/Editing/class.EditingGUIRequest.php [hunk 7]
```diff
@@ -24,7 +24,6 @@ class EditingGUIRequest
 {
     use BaseGUIRequest;
 
-    protected \ILIAS\HTTP\Services $http;
```
Comment (mjansenDatabay) [inline Modules/Survey/Editing/class.EditingGUIRequest.php:27]: This is already provided by the trait

File: Modules/Survey/Execution/class.ExecutionGUIRequest.php [hunk 8]
```diff
@@ -24,7 +24,6 @@ class ExecutionGUIRequest
 {
     use BaseGUIRequest;
 
-    protected \ILIAS\HTTP\Services $http;
```
Comment (mjansenDatabay) [inline Modules/Survey/Execution/class.ExecutionGUIRequest.php:27]: This is already provided by the trait

File: Modules/Survey/Execution/class.ilSurveyExecutionGUI.php [hunk 9]
```diff
@@ -298,7 +298,7 @@ public function previous(
         }
         $this->ctrl->setParameter($this, "activecommand", "previous");
         $this->ctrl->setParameter($this, "qid", $this->request->getQuestionId());
-        if (strlen($has_error)) {
+        if (strlen($has_error)) {// TODO: strlen MUST be used with strings, saveUserInput returns an int / A string cast will always evaluate in min. 1 character
```
Comment (mjansenDatabay) [inline Modules/Survey/Execution/class.ilSurveyExecutionGUI.php:301]: Please have a look here

File: Modules/Survey/Execution/class.ilSurveyExecutionGUI.php [hunk 10]
```diff
@@ -314,7 +314,7 @@ public function next() : void
         $result = $this->saveUserInput("next");
         $this->ctrl->setParameter($this, "activecommand", "next");
         $this->ctrl->setParameter($this, "qid", $this->request->getQuestionId());
-        if (strlen($result)) {
+        if (strlen($result)) {// TODO: strlen MUST be used with strings, saveUserInput returns an int / A string cast will always evaluate in min. 1 character
```
Comment (mjansenDatabay) [inline Modules/Survey/Execution/class.ilSurveyExecutionGUI.php:317]: Please have a look here

File: Modules/Survey/Editing/class.ilSurveyEditorGUI.php [hunk 11]
```diff
@@ -639,7 +639,7 @@ public function createQuestionObject(
         return null;
     }
 
-    public function executeCreateQuestionObject($q_type = null, $pool_usage = null, $pgov_pos = null) : void
+    public function executeCreateQuestionObject($q_type = null, $pool_usage = null, $pgov_pos = null) : void// TODO PHP8-REVIEW Missing types
```
Comment (mjansenDatabay) [inline Modules/Survey/Editing/class.ilSurveyEditorGUI.php:642]: Types are missing here on appropriate `PHPDoc` comments

File: Modules/Survey/Editing/class.ilSurveyEditorGUI.php [hunk 12]
```diff
@@ -663,11 +663,11 @@ public function executeCreateQuestionObject($q_type = null, $pool_usage = null,
             $obj_id = $this->object->getId();
         }
         // existing pool
-        elseif ($pool_usage == 3 && strlen($this->request->getSelectedPool())) {
+        elseif ($pool_usage == 3 && strlen($this->request->getSelectedPool())) {// TODO PHP8-REVIEW strlen() vs. int
```
Comment (mjansenDatabay) [inline Modules/Survey/Editing/class.ilSurveyEditorGUI.php:666]: `strlen()` expects a `string`, an `int` is provided here (may be also `0`)

File: Modules/Survey/Editing/class.ilSurveyEditorGUI.php [hunk 13]
```diff
@@ -663,11 +663,11 @@ public function executeCreateQuestionObject($q_type = null, $pool_usage = null,
...
             $obj_id = ilObject::_lookupObjId($this->request->getSelectedPool());
         }
         // new pool
-        elseif ($pool_usage == 2 && strlen($this->request->getPoolName())) {
+        elseif ($pool_usage == 2 && strlen($this->request->getPoolName())) {// TODO PHP8-REVIEW strlen() vs. int
```
Comment (mjansenDatabay) [inline Modules/Survey/Editing/class.ilSurveyEditorGUI.php:670]: `strlen()` expects a `string`, an `int` is provided here (may be also `0`)

File: Modules/Survey/Execution/class.ilSurveyExecutionGUI.php [hunk 14]
```diff
@@ -298,7 +298,7 @@ public function previous(
         }
         $this->ctrl->setParameter($this, "activecommand", "previous");
         $this->ctrl->setParameter($this, "qid", $this->request->getQuestionId());
-        if (strlen($has_error)) {
+        if (strlen($has_error)) {// TODO PHP8-REVIEW strlen MUST be used with strings, saveUserInput returns an int / A string cast will always evaluate in min. 1 character
```
Comment (mjansenDatabay) [inline Modules/Survey/Execution/class.ilSurveyExecutionGUI.php:301]: `strlen()` expects a `string`, an `int` is provided here (may be also `0`)

File: Modules/Survey/Execution/class.ilSurveyExecutionGUI.php [hunk 15]
```diff
@@ -314,7 +314,7 @@ public function next() : void
         $result = $this->saveUserInput("next");
         $this->ctrl->setParameter($this, "activecommand", "next");
         $this->ctrl->setParameter($this, "qid", $this->request->getQuestionId());
-        if (strlen($result)) {
+        if (strlen($result)) {// TODO PHP8-REVIEW strlen MUST be used with strings, saveUserInput returns an int / A string cast will always evaluate in min. 1 character
```
Comment (mjansenDatabay) [inline Modules/Survey/Execution/class.ilSurveyExecutionGUI.php:317]: `strlen()` expects a `string`, an `int` is provided here (may be also `0`)

File: Modules/Survey/Export/class.ilSurveyDataSet.php [hunk 16]
```diff
@@ -130,15 +130,15 @@ public function importRecord(
                 $skill_data = ilBasicSkill::getCommonSkillIdForImportId($this->getCurrentInstallationId(), $a_rec["BaseSkillId"], $a_rec["TrefId"]);
                 $q_id = $a_mapping->getMapping("Modules/Survey", "svy_q", $a_rec["QId"]);
                 if ($q_id > 0 && count($skill_data) > 0) {
-                    $skill_survey = new ilSurveySkill($this->getImport()->getSurvey());
+                    $skill_survey = new ilSurveySkill($this->getImport()->getSurvey());// TODO PHP8-REVIEW No method `getSurvey`
```
Comment (mjansenDatabay) [inline Modules/Survey/Export/class.ilSurveyDataSet.php:133]: Method `getSurvey()` does not exist

File: Modules/Survey/Export/class.ilSurveyDataSet.php [hunk 17]
```diff
@@ -130,15 +130,15 @@ public function importRecord(
...
                     $skill_survey->addQuestionSkillAssignment($q_id, $skill_data[0]["skill_id"], $skill_data[0]["tref_id"]);
                 }
                 break;
 
             case "svy_skill_threshold":
                 $l = ilBasicSkill::getLevelIdForImportIdMatchSkill($this->getCurrentInstallationId(), $a_rec["LevelId"], $a_rec["BaseSkillId"], $a_rec["TrefId"]);
                 if (count($l) > 0) {
-                    $skill_thres = new ilSurveySkillThresholds($this->getImport()->getSurvey());
+                    $skill_thres = new ilSurveySkillThresholds($this->getImport()->getSurvey());// TODO PHP8-REVIEW No method `getSurvey`
```
Comment (mjansenDatabay) [inline Modules/Survey/Export/class.ilSurveyDataSet.php:141]: Method `getSurvey()` does not exist

File: Modules/Survey/Export/class.ilSurveyImporter.php [hunk 18]
```diff
@@ -43,7 +43,7 @@ public function init() : void
     {
         $this->ds = new ilSurveyDataSet();
         $this->ds->setDSPrefix("ds");
-        $this->ds->setImport($this);
+        $this->ds->setImport($this);// TODO PHP8-REVIEW Wrong type
```
Comment (mjansenDatabay) [inline Modules/Survey/Export/class.ilSurveyImporter.php:46]: The passed type is wrong

File: Modules/Survey/Participants/class.ilSpecialUsersTableGUI.php [hunk 19]
```diff
@@ -46,7 +46,7 @@ public function __construct(
     protected function fillRow(array $a_set) : void
     {
         $ilCtrl = $this->ctrl;
-        $user = ilObjUser::_lookupFields($a_set);
+        $user = ilObjUser::_lookupFields($a_set);// TODO PHP8-REVIEW array passed, int expected
```
Comment (mjansenDatabay) [inline Modules/Survey/Participants/class.ilSpecialUsersTableGUI.php:49]: The passed type is wrong

File: Modules/Survey/Skills/class.ilSurveySkillGUI.php [hunk 20]
```diff
@@ -95,7 +95,7 @@ public function assignSkillToQuestion() : void
         $ilCtrl = $this->ctrl;
         $ilCtrl->saveParameter($this, "q_id");
 
-        $sel = new ilSkillSelectorGUI($this, "assignSkillToQuestion", $this, "selectSkillForQuestion");
+        $sel = new ilSkillSelectorGUI($this, "assignSkillToQuestion", $this, "selectSkillForQuestion", '');
```
Comment (mjansenDatabay) [inline Modules/Survey/Skills/class.ilSurveySkillGUI.php]: Wrong types are used here

File: Modules/Survey/Participants/class.ilSurveyParticipantsTableGUI.php [hunk 21]
```diff
@@ -1,4 +1,4 @@
-<?php
+ <?php
```
Comment (alex40724) [inline Modules/Survey/Participants/class.ilSurveyParticipantsTableGUI.php:1]: @mjansenDatabay ?

Comment (mjansenDatabay) [reply Modules/Survey/Participants/class.ilSurveyParticipantsTableGUI.php:1]: Hmmm, no idea how this occured. Maybe this is not corrected by the `PHP-CS-FIXER`.

**Comment by alex40724** [comment]
Thanks for all the changes. I will work through the todos now...

**Comment by alex40724** [comment]
done.

**Pull request merged**